src/main/java/com/mongodb/hibernate/translate/attachment/Attachment.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2024-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.hibernate.translate.attachment;
+
+import static java.lang.String.format;
+
+import com.mongodb.hibernate.internal.VisibleForTesting;
+import org.jspecify.annotations.Nullable;
+
+/**
+ * Manages global state while translating SQL AST via {@link org.hibernate.sql.ast.SqlAstWalker}.
+ *
+ * <p>{@code SqlAstWalker} is based on {@code visitor} pattern without returning value from the various visitor methods.
+ * The original goal of {@code SqlAstWalker} is for concatenating SQL string internally; but when translating SQL AST
+ * into MQL, it is highly desirable to return value (e.g. {@link com.mongodb.hibernate.translate.mongoast.AstNode}). To
+ * avoid such dilemma, some attachment/detachment mechanism is implemented as follows:
+ *
+ * <ul>
+ *   <li>when some visitor method needs to return some value, it attaches the value by invoking
+ *       {@link #attach(AttachmentKey, Object)}.
+ *   <li>when some coordinating detacher needs to access the value attached above, it invokes
+ *       {@link #expect(AttachmentKey, Runnable)}; internally the {@code Runnable} will attach value by the above
+ *       attacher.
+ *   <li>both attacher and detacher should provide identical {@link AttachmentKey} to ensure data transferring safety.
+ * </ul>
+ *
+ * @see org.hibernate.sql.ast.SqlAstWalker
+ * @see AttachmentKey
+ */
+@SuppressWarnings("unchecked")
+public final class Attachment {
+
+    private @Nullable AttachmentKey<?> attachmentKey;
+    private @Nullable Object attachedValue;
+
+    @Override
+    public String toString() {
+        return "Attachment{" + "attachmentKey=" + attachmentKey + ", attachedValue=" + attachedValue + "}";
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.AccessModifier.PRIVATE)
+    boolean isBlank() {
+        return this.attachmentKey == null && this.attachedValue == null;
+    }
+
+    /**
+     * Detaches the attached value by invoking some visitor method which returns {@code void} but internally the method
```

Comment:
The "detaches" part (`this.attachedValue = null`) is conceptually an implementation detail, it cannot in principle be part of the method contract because it is not an observable behavior, provided that the `expect` methods is not called concurrently with itself. Thus, it should not be mentioned or even alluded to in the documentation.